File: mos/backend/app/schemas/funnels.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal, Optional

from pydantic import BaseModel, Field


class FunnelCreateRequest(BaseModel):
    clientId: str
    campaignId: Optional[str] = None
    experimentId: Optional[str] = None
    productId: Optional[str] = None
    selectedOfferId: Optional[str] = None
    name: str
    description: Optional[str] = None


class FunnelUpdateRequest(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    campaignId: Optional[str] = None
    experimentId: Optional[str] = None
    entryPageId: Optional[str] = None
    designSystemId: Optional[str] = None
    productId: Optional[str] = None
    selectedOfferId: Optional[str] = None


class FunnelDuplicateRequest(BaseModel):
    targetCampaignId: Optional[str] = None
    name: Optional[str] = None
    copyMode: Literal["approvedOnly", "activePublication"] = "approvedOnly"
    autoPublish: bool = False


class FunnelPageCreateRequest(BaseModel):
    name: str
    slug: Optional[str] = None
    templateId: Optional[str] = None
    designSystemId: Optional[str] = None
    nextPageId: Optional[str] = None


class FunnelPageUpdateRequest(BaseModel):
    name: Optional[str] = None
    slug: Optional[str] = None
    ordering: Optional[int] = None
    designSystemId: Optional[str] = None
    nextPageId: Optional[str] = None


class FunnelPageSaveDraftRequest(BaseModel):
    puckData: dict[str, Any]


class PublicFunnelMetaResponse(BaseModel):
    publicId: str
    funnelId: str
    publicationId: str
    entrySlug: str
    pages: list[dict[str, str]]


class PublicFunnelPageResponse(BaseModel):
    funnelId: str
    publicationId: str
    pageId: str
    slug: str
    puckData: dict[str, Any]
    pageMap: dict[str, str]
    designSystemTokens: Optional[dict[str, Any]] = None
    nextPageId: Optional[str] = None


class PublicFunnelRedirectResponse(BaseModel):
    redirectToSlug: str


class PublicEventIn(BaseModel):
    eventType: str
    occurredAt: Optional[datetime] = None
    publicationId: str
    pageId: str
    visitorId: Optional[str] = None
    sessionId: Optional[str] = None
    path: Optional[str] = None
    referrer: Optional[str] = None
    utm: dict[str, Any] = Field(default_factory=dict)
    props: dict[str, Any] = Field(default_factory=dict)


class PublicEventsIngestRequest(BaseModel):
    events: list[PublicEventIn]


class GenerateFunnelImageRequest(BaseModel):
    prompt: str
    clientId: str
    aspectRatio: Optional[str] = None
    styleHints: Optional[dict[str, Any]] = None
    usageContext: Optional[dict[str, Any]] = None


class GenerateFunnelImageResponse(BaseModel):
    assetId: str
    publicId: str
    width: Optional[int] = None
    height: Optional[int] = None


class FunnelAIChatMessage(BaseModel):
    role: Literal["user", "assistant"]
    content: str


class FunnelAIAttachment(BaseModel):
    assetId: str
    publicId: str
    filename: Optional[str] = None
    contentType: Optional[str] = None
    width: Optional[int] = None
    height: Optional[int] = None


class FunnelPageAIGenerateRequest(BaseModel):
    prompt: str
    messages: list[FunnelAIChatMessage] = Field(default_factory=list)
    attachedAssets: list[FunnelAIAttachment] = Field(default_factory=list)
    copyPack: Optional[str] = None
    currentPuckData: Optional[dict[str, Any]] = None
    templateId: Optional[str] = None
    ideaWorkspaceId: Optional[str] = None
    model: Optional[str] = None
    temperature: float = 0.2
    maxTokens: Optional[int] = None
    generateImages: bool = True
    maxImages: int = 3


class FunnelPageAIGenerateResponse(BaseModel):
    assistantMessage: str
    puckData: dict[str, Any]
    draftVersionId: str
    generatedImages: list[dict[str, Any]] = Field(default_factory=list)
    imagePlans: list[dict[str, Any]] = Field(default_factory=list)


class FunnelPageTestimonialGenerateRequest(BaseModel):
    draftVersionId: Optional[str] = None
    currentPuckData: Optional[dict[str, Any]] = None
    templateId: Optional[str] = None
    ideaWorkspaceId: Optional[str] = None
    model: Optional[str] = None
    temperature: float = 0.3
    maxTokens: Optional[int] = None
    synthetic: bool = True


class FunnelPageTestimonialGenerateResponse(BaseModel):
    draftVersionId: str
    puckData: dict[str, Any]
    generatedTestimonials: list[dict[str, Any]] = Field(default_factory=list)


class FunnelTemplateSummary(BaseModel):
    id: str
    name: str
    description: Optional[str] = None
    previewImage: Optional[str] = None


class FunnelTemplateDetail(FunnelTemplateSummary):
    puckData: dict[str, Any]

File: mos/frontend/src/types/funnels.ts
export type FunnelStatus = "draft" | "published" | "disabled" | "archived";

import type { DesignSystemTokens } from "@/types/designSystems";
import type { MedusaRuntimeConfig } from "@/lib/medusa/config";

export type Funnel = {
  id: string;
  org_id: string;
  client_id: string;
  product_id?: string | null;
  selected_offer_id?: string | null;
  campaign_id: string | null;
  experiment_spec_id?: string | null;
  design_system_id?: string | null;
  name: string;
  description: string | null;
  status: FunnelStatus;
  route_slug: string;
  public_id: string;
  entry_page_id: string | null;
  active_publication_id: string | null;
  created_at: string;
  updated_at: string;
};

export type FunnelPage = {
  id: string;
  funnel_id: string;
  name: string;
  slug: string;
  next_page_id?: string | null;
  template_id?: string | null;
  design_system_id?: string | null;
  ordering: number;
  created_at: string;
  updated_at: string;
  latestDraftVersionId?: string | null;
  latestApprovedVersionId?: string | null;
};

export type FunnelDetail = Funnel & {
  pages: FunnelPage[];
  canPublish: boolean;
};

export type FunnelPageVersionStatus = "draft" | "approved";

export type FunnelPageVersion = {
  id: string;
  page_id: string;
  status: FunnelPageVersionStatus;
  puck_data: unknown;
  source: string;
  ai_metadata: unknown | null;
  created_at: string;
};

export type FunnelPageDetail = {
  page: FunnelPage;
  latestDraft: FunnelPageVersion | null;
  latestApproved: FunnelPageVersion | null;
  designSystemTokens?: DesignSystemTokens | null;
};

export type PublicFunnelMeta = {
  productSlug: string;
  funnelSlug: string;
  funnelId: string;
  publicationId: string;
  entrySlug: string;
  pages: { pageId: string; slug: string }[];
  medusaRuntimeConfig?: MedusaRuntimeConfig | null;
};

export type PublicFunnelStage = "pre_sales" | "sales" | "checkout" | "thank_you" | "custom";

export type HtmlDeployArtifactKind = "listicle" | "listicle_hybrid" | "quiz" | "sales";

export type ImportedHtmlTrackEventType =
  | "pre_sales_to_sales_click"
  | "sales_to_checkout_click"
  | "checkout_started"
  | "selector_interaction"
  | "product_detail_interaction"
  | "custom_page_click";

export type ImportedHtmlOptionSelector = {
  name: string;
  selector: string;
  source: "value" | "text";
};

export type ImportedHtmlVariantResolver =
  | {
      type: "fixed";
      variantId: string;
    }
  | {
      type: "option_values";
      optionSelectors: ImportedHtmlOptionSelector[];
    };

export type ImportedHtmlCheckoutConfig =
  | {
      mode: "public_checkout";
      variantResolver: ImportedHtmlVariantResolver;
    }
  | {
      mode: "external_checkout_url";
      variantResolver: ImportedHtmlVariantResolver;
      externalUrlsByVariant: Array<{
        variantId: string;
        url: string;
      }>;
    };

export type ImportedHtmlInstrumentationBinding =
  | {
      id: string;
      type: "internal_navigation";
      selector: string;
      event: "click";
      targetPageId: string;
      trackEventType: ImportedHtmlTrackEventType;
    }
  | {
      id: string;
      type: "checkout";
      selector: string;
      event: "click";
      trackEventType: ImportedHtmlTrackEventType;
      checkout: ImportedHtmlCheckoutConfig;
    }
  | {
      id: string;
      type: "track_only";
      selector: string;
      event: "click";
      trackEventType: ImportedHtmlTrackEventType;
    };

export type ImportedHtmlViewTarget = {
  id: string;
  selector: string;
  label?: string | null;
  proofType?: string | null;
  sectionId?: string | null;
  ctaPosition?: number | null;
  questionId?: string | null;
  questionText?: string | null;
  questionIndex?: number | null;
  questionType?: string | null;
  questionRole?: string | null;
  isRequired?: boolean | null;
  optionId?: string | null;
  optionText?: string | null;
  optionIndex?: number | null;
  optionPosition?: number | null;
  optionRole?: string | null;
  selectionOrder?: number | null;
  submitOnSelect?: boolean | null;
  resultId?: string | null;
  segmentId?: string | null;
  screenIndex?: number | null;
  screenName?: string | null;
  route?: string | null;
  hash?: string | null;
  titleContains?: string | null;
  requiredBreakpoints?: number[];
  recommendationId?: string | null;
  screenIndex?: number | null;
  screenName?: string | null;
  route?: string | null;
  hash?: string | null;
  titleContains?: string | null;
  requiredBreakpoints?: number[];
  offerId?: string | null;
  sku?: string | null;
  mechanismName?: string | null;
  guaranteeType?: string | null;
  interactionType?: string | null;
  selectedValue?: string | null;
  event?: "click" | "change" | "input" | null;
  source?: "value" | "text" | "checked" | null;
  quizId?: string | null;
  quizVersion?: string | null;
  quizVariant?: string | null;
  answerPathId?: string | null;
  angle?: string | null;
  awarenessLevel?: string | null;
  sophisticationLevel?: string | null;
  angleFamily?: string | null;
  hookId?: string | null;
  promiseId?: string | null;
  bundleId?: string | null;
  pricePoint?: string | null;
  guaranteeId?: string | null;
  guaranteeDuration?: string | null;
  valueTotal?: number | null;
  actualPrice?: number | null;
  valueRatio?: number | null;
  clickType?: string | null;
  targetOfferId?: string | null;
  destinationUrl?: string | null;
  elementId?: string | null;
  subscriptionFlag?: boolean | null;
};

export type ImportedHtmlInstrumentationManifest = {
  schemaVersion: "html-deploy-v1";
  htmlArtifactKind: HtmlDeployArtifactKind;
  pageStage: PublicFunnelStage;
  quizId?: string | null;
  quizVersion?: string | null;
  quizVariant?: string | null;
  bindings: ImportedHtmlInstrumentationBinding[];
  sections?: ImportedHtmlViewTarget[];
  proofs?: ImportedHtmlViewTarget[];
  ctas?: ImportedHtmlViewTarget[];
  offerStacks?: ImportedHtmlViewTarget[];
  valueStacks?: ImportedHtmlViewTarget[];
  priceReveals?: ImportedHtmlViewTarget[];
  guarantees?: ImportedHtmlViewTarget[];
  trustElements?: ImportedHtmlViewTarget[];
  quizLeads?: ImportedHtmlViewTarget[];
  quizQuestions?: ImportedHtmlViewTarget[];
  quizOptions?: ImportedHtmlViewTarget[];
  quizSubmissions?: ImportedHtmlViewTarget[];
  quizScrollTargets?: ImportedHtmlViewTarget[];
  quizResults?: ImportedHtmlViewTarget[];
  quizMechanisms?: ImportedHtmlViewTarget[];
  quizRecommendations?: ImportedHtmlViewTarget[];
  quizScrollTargets?: ImportedHtmlViewTarget[];
  productDetails?: ImportedHtmlViewTarget[];
  selectors?: ImportedHtmlViewTarget[];
};

// Site page types for commerce experiences
export type SitePageType =
  | "home"
  | "store"
  | "collection"
  | "category"
  | "product_detail"
  | "cart"
  | "checkout"
  | "privacy_policy"
  | "terms_of_service"
  | "returns_refunds_policy"
  | "shipping_policy"
  | "contact_support"
  | "account_dashboard"
  | "account_profile"
  | "account_addresses"
  | "account_orders"
  | "account_order_detail"
  | "order_confirmed"
  | "order_transfer"
  | "order_transfer_accept"
  | "order_transfer_decline";

export type PublicFunnelPage = {
  productSlug: string;
  funnelId: string;
  publicationId: string;
  pageId: string;
  slug: string;
  stage: PublicFunnelStage;
  puckData: unknown;
  pageMap: Record<string, string>;
  pageStageMap: Record<string, PublicFunnelStage>;
  pageTypeMap?: Record<string, SitePageType>;
  designSystemTokens?: DesignSystemTokens | null;
  metadata?: {
    title: string;
    description: string;
    lang: string;
    brandName?: string | null;
  };
  tracking?: {
    provider: string;
    mode: string;
    metaPixelId?: string | null;
    posthogProjectApiKey?: string | null;
    posthogApiHost?: string | null;
    posthogUiHost?: string | null;
    posthogDefaults?: string | null;
    posthogPersonProfiles?: "identified_only" | "always" | null;
  } | null;
  nextPageId?: string | null;
  redirectToSlug?: string;
};

export type FunnelImageAsset = {
  assetId: string;
  publicId: string;
  width: number | null;
  height: number | null;
  url: string;
};

export type FunnelAIChatMessage = {
  role: "user" | "assistant";
  content: string;
};

export type FunnelAIAttachment = {
  assetId: string;
  publicId: string;
  filename?: string | null;
  contentType?: string | null;
  width?: number | null;
  height?: number | null;
};

export type FunnelPageAIGenerateResponse = {
  assistantMessage: string;
  puckData: unknown;
  draftVersionId: string;
  generatedImages: Array<Record<string, unknown>>;
  imagePlans: Array<Record<string, unknown>>;
};

export type FunnelTemplateSummary = {
  id: string;
  name: string;
  description?: string | null;
  previewImage?: string | null;
  category?: string | null;
};

export type FunnelTemplateDetail = FunnelTemplateSummary & {
  puckData: unknown;
};
